include/atomics/sender_cadmium.hpp

```diff
@@ -0,0 +1,183 @@
+
+#include "../data_structures/message.hpp"
+
+using namespace cadmium;
+using namespace std;
+
+//Port definition
+struct Sender_defs {
+	struct packetSentOut: public out_port<Message_t> {
+	};
+	struct ackReceivedOut: public out_port<Message_t> {
+	};
+	struct dataOut: public out_port<Message_t> {
+	};
+	struct controlIn: public in_port<Message_t> {
+	};
+	struct ackIn: public in_port<Message_t> {
+	};
+};
+
+template<typename TIME>
+class Sender {
+	using defs=Sender_defs; // putting definitions in context
+public:
+	//Parameters to be overwriten when instantiating the atomic model
+	TIME preparation_time;
+	TIME time_out;
+	// default constructor
+	Sender() noexcept {
+		preparation_time = TIME("00:00:10");
+		time_out = TIME("00:00:20");
+		state.alt_bit = 0;
+		state.next_internal = std::numeric_limits<TIME>::infinity();
+		state.model_active = false;
+	}
+
+	// state definition
+	struct state_type {
+		bool ack;
+		int packet_num;
+		int total_packet_num;
+		int alt_bit;
+		bool sending;
+		bool model_active;
+		TIME next_internal;
+	};
+	state_type state;
+	// ports definition
+	using input_ports=std::tuple<typename defs::controlIn, typename defs::ackIn>;
+	using output_ports=std::tuple<typename defs::packetSentOut,
+			typename defs::ackReceivedOut, typename defs::dataOut>;
+
+	// internal transition
+	void internal_transition() {
+		if (state.ack) {
+			if (state.packet_num < state.total_packet_num) {
+				state.packet_num++;
+				state.ack = false;
+				state.alt_bit = (state.alt_bit + 1) % 2;
+				state.sending = true;
+				state.model_active = true;
+				state.next_internal = preparation_time;
+			} else {
+				state.model_active = false;
+				state.next_internal = std::numeric_limits<TIME>::infinity();
+			}
+		} else {
+			if (state.sending) {
+				state.sending = false;
+				state.model_active = true;
+				state.next_internal = time_out;
+			} else {
+				state.sending = true;
+				state.model_active = true;
+				state.next_internal = preparation_time;
+			}
+		}
+	}
+
+	// external transition
+	void external_transition(TIME e,
+			typename make_message_bags<input_ports>::type mbs) {
+		if ((get_messages<typename defs::controlIn>(mbs).size()
+				+ get_messages<typename defs::ackIn>(mbs).size()) > 1)
+			assert(false && "one message per time uniti");
+		for (const auto &x : get_messages<typename defs::controlIn>(mbs)) {
+			if (state.model_active == false) {
+				state.total_packet_num = static_cast<int>(x.value);
+				if (state.total_packet_num > 0) {
+					state.packet_num = 1;
+					state.ack = false;
+					state.sending = true;
+					state.alt_bit = state.packet_num % 2;  //set initial alt_bit
+					state.model_active = true;
+					state.next_internal = preparation_time;
+				} else {
+					if (state.next_internal
+							!= std::numeric_limits<TIME>::infinity()) {
+						state.next_internal = state.next_internal - e;
+					}
+				}
+			}
+		}
+		for (const auto &x : get_messages<typename defs::ackIn>(mbs)) {
```

Comment:
Check for indentation

Reply:
done.. please open this in an editor and check